lib/packwerk/deprecated_references.rb

```diff
@@ -36,28 +36,41 @@ def listed?(reference, violation_type:)
     sig { params(reference: Packwerk::Reference, violation_type: Packwerk::ViolationType).returns(T::Boolean) }
     def add_entries(reference, violation_type)
       package_violations = @new_entries.fetch(reference.constant.package.name, {})
-      entries_for_file = package_violations[reference.constant.name] ||= {}
+      entries_for_constant = package_violations[reference.constant.name] ||= {}
 
-      entries_for_file["violations"] ||= []
-      entries_for_file["violations"] << violation_type.serialize
+      entries_for_constant["violations"] ||= []
+      entries_for_constant["violations"] << violation_type.serialize
 
-      entries_for_file["files"] ||= []
-      entries_for_file["files"] << reference.relative_path.to_s
+      entries_for_constant["files"] ||= []
+      entries_for_constant["files"] << reference.relative_path.to_s
 
       @new_entries[reference.constant.package.name] = package_violations
       listed?(reference, violation_type: violation_type)
     end
 
-    sig { returns(T::Boolean) }
-    def stale_violations?
+    sig { params(for_files: T::Set[String]).returns(T::Boolean) }
+    def stale_violations?(for_files)
       prepare_entries_for_dump
+
       deprecated_references.any? do |package, package_violations|
-        package_violations.any? do |constant_name, entries_for_file|
+        package_violations_for_files = {}
+        package_violations.each do |constant_name, entries_for_constant|
+          entries_for_files = for_files & entries_for_constant["files"]
+          next if entries_for_files.none?
+          package_violations_for_files[constant_name] = {
+            "violations" => entries_for_constant["violations"],
+            "files" => entries_for_files.to_a,
+          }
+        end
```

Comment:
What do you think of extracting this filtering step into its own method?

Is there much performance impact when running stale violations checks on the entire codebase as a result of essentially copying all the violations data?

Reply:
I ran this against our codebase and didn't notice any huge changes in performance. I'm pretty sure the variation is just noise.

![Uploading Screen Shot 2022-05-10 at 2.03.37 PM.png…]()